parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java

```diff
@@ -0,0 +1,799 @@
+  @Override
+  public void addLong(long value) {
+    builder.builder.appendTimestampNtz(value);
+  }
+}
+
+class VariantTimestampNanosConverter extends VariantScalarConverter {
+  @Override
+  public void addLong(long value) {
+    builder.builder.appendTimestampNanos(value);
+  }
+}
+
+class VariantTimestampNanosNtzConverter extends VariantScalarConverter {
+  @Override
+  public void addLong(long value) {
+    builder.builder.appendTimestampNanosNtz(value);
+  }
+}
+
+/**
+ * Converter for a LIST typed_value.
+ */
+class VariantArrayConverter extends GroupConverter implements VariantConverter {
+  private VariantBuilderHolder builder;
+  private VariantArrayRepeatedConverter repeatedConverter;
+  private ArrayList<Integer> offsets;
+  private int startPos;
+
+  public VariantArrayConverter(GroupType listType) {
+    if (listType.getFieldCount() != 1) {
+      throw new IllegalArgumentException("LIST must have one field");
+    }
+    Type middleLevel = listType.getType(0);
+    if (!middleLevel.isRepetition(REPEATED)
+        || middleLevel.isPrimitive()
+        || middleLevel.asGroupType().getFieldCount() != 1) {
+      throw new IllegalArgumentException("LIST must have one repeated field");
+    }
+    this.repeatedConverter = new VariantArrayRepeatedConverter(middleLevel.asGroupType(), this);
+  }
+
+  @Override
+  public void init(VariantBuilderHolder builderHolder) {
+    builder = builderHolder;
+    repeatedConverter.init(builderHolder);
+  }
+
+  @Override
+  public Converter getConverter(int fieldIndex) {
+    return repeatedConverter;
+  }
+
+  public void addElement() {
+    offsets.add(builder.builder.getWritePos() - startPos);
+  }
+
+  @Override
+  public void start() {
+    offsets = new ArrayList<>();
+    startPos = builder.builder.getWritePos();
+  }
+
+  @Override
+  public void end() {
+    builder.builder.finishWritingArray(startPos, offsets);
+  }
+}
+
+/**
+ * Converter for the repeated field of a LIST typed_value.
+ */
+class VariantArrayRepeatedConverter extends GroupConverter implements VariantConverter {
+  private VariantElementConverter elementConverter;
+  private VariantArrayConverter parentConverter;
+
+  public VariantArrayRepeatedConverter(GroupType repeatedType, VariantArrayConverter parentaConverter) {
+    this.elementConverter = new VariantElementConverter(repeatedType.getType(0).asGroupType());
+    this.parentConverter = parentaConverter;
+  }
+
+  @Override
+  public void init(VariantBuilderHolder builderHolder) {
+    elementConverter.init(builderHolder);
+  }
+
+  @Override
+  public Converter getConverter(int fieldIndex) {
+    return elementConverter;
+  }
+
+  @Override
+  public void start() {
+    // Record the offset of this element in the binary.
+    parentConverter.addElement();
+  }
+
+  @Override
+  public void end() {}
+}
+
+class VariantObjectConverter extends GroupConverter implements VariantConverter {
+  private VariantBuilderHolder builder;
```

Comment:
Calling this `builder` results in later calls to `builder.builder.getWritePos()`. It would be better to name this `holder`.